### ISSUE TITLE
Environment variable not set from a dependency of a task that sets `watch` to true (plus, the dependency is executed twice)

### ISSUE THREAD

**Issue by d4h0** [sagiegurari/cargo-make#515]
### Describe The Bug

Hi,

It seems, there is a problem when setting environment variables from a dependency of a task that sets `watch` to true.

### To Reproduce

1) create a new Rust project via `cargo init xy`
2) put the following content in the Makefile.toml:

```
# Clears the terminal screen. Skips the first time to not clear the screen after `start-dev-server` is started when `dev-flow` i
s executed
[tasks.clear-terminal-skip-first]
script_runner = "@duckscript"
script = '''
skip_1 = get_env CLEAR_TERMINAL_SKIP_1
skip_2 = get_env CLEAR_TERMINAL_SKIP_2
echo skip_1 = ${skip_1}
echo skip_2 = ${skip_2}
if is_empty ${skip_1}
    echo FIRST
    set_env CLEAR_TERMINAL_SKIP_1 true # done
elseif is_empty ${skip_2}
    # For some reason, this script is executed twice for the first execution of "watch-run" within "dev-flow" (I guess, because of the `watch` option of "watch-run"). Therfore we actually have to skip two executions
    echo SECOND
    set_env CLEAR_TERMINAL_SKIP_2 true # done
    set_env CLEAR_TERMINAL_SKIP_1 false # done
else
    echo THEN
    #cm_run_task "clear-terminal"
    #exec clear
end
'''

[tasks.watch-run]
command = "cargo"
args = ["run"]
watch = true
dependencies = ["clear-terminal-skip-first"]

[tasks.dev-flow]
run_task = {name = [
    #"start-dev-server",
    "watch-run",
], parallel = true}
```
3) run cargo make dev-flow

4) execute `touch src/main.rs`

### Error Stack

No error, but as you can see, the env variable `CLEAR_TERMINAL_SKIP_2` is not set. If I didn't make a mistake, the above task should work, right?

I also didn't expect, that "clear-terminal-skip-first" is executed twice for a single task run (see "clear-terminal-skip-first" and output).

Is that a bug?

**Comment by sagiegurari** [comment]
thanks for reporting. I'll check it out :)

**Comment by d4h0** [comment]
Okay, thank you

Btw., is there an easier / a different way to run a dependency of a task that has activated `watch` only after the first run?

**Comment by sagiegurari** [comment]
feel like you are looking for something like this:

```toml
[tasks.clear-terminal-skip-first]
# .....

[tasks.run]
command = "cargo"
args = ["run"]

[tasks.watch-run]
run_task = { name = ["clear-terminal", "run"] }
watch = true
```

which will every iteration run clear and run.

**Comment by d4h0** [comment]
No, I don't think this would do what I intend.

What I'd like to accomplish is this:

1) start a dev server, which prints the address it's listening to

2) start cargo watch and compile my project

3) Don't clear the terminal at this point, so I can see the address of the dev server

4) if I make a change, the terminal is cleared and cargo run is executed (via cargo watch)

If I add `postpone: true` to `cargo watch` (which wouldn't clear the termninal), I always have to change a file, so my project is compiled for the first time. This is not ideal, because my project is a static website generator, and my website assets (HTML, CSS, templates, etc.) are excluded from `cargo watch` (my project is watching these files itself. No re-compilation is needed here). If I work on the design of my website, for example, I have to make a dummy change to a Rust file to trigger `cargo run` via `cargo watch`.

So I have the option to disable `postpone` and not see the address, or to make dummy changes, which isn't a great developer experience. Hence, I tried to create a `terminal clear` task, which skips the initial `cargo run` (which would be ideal).

**Comment by sagiegurari** [comment]
thanks for the detailed flow, that helps. so how about this?

```toml
[tasks.clear-terminal]
condition = { env_true = ["CLEAR_TERMINAL"] }
# .....

[tasks.enable-clear-terminal]
script_runner = "@duckscript"
script = '''
set_env CLEAR_TERMINAL true
'''

[tasks.run]
command = "cargo"
args = ["run"]

[tasks.watch-run]
run_task = { name = ["clear-terminal", "enable-clear-terminal", "run"] }
watch = true
```

**Comment by d4h0** [comment]
Thanks, @sagiegurari!

Unfortunately, this doesn't work... :)

(the terminal isn't cleared at all)

If you run my flow from the initial comment, can you reproduce what I describe?

I think, there might be a bug (or even several). If you look at my code and the output it produces, it should be clearer what I'm talking about.

I assume, that `cargo watch` is executed in a different context, and that environment variables are not synced with `cargo make` (but I might be wrong).

**Comment by sagiegurari** [comment]
oh, didn't think about it. let me check it out and see if i can help resolve this.

**Comment by d4h0** [comment]
Thank you for your help :)

**Comment by sagiegurari** [comment]
haven't tried it, but i think something like this should work.
basically since we can't pass env vars from child to parent process (thanks for reminding me that watch forks it), i'm using a file as a marker.

```toml
[env]
CLEAR_TERMINAL_MARKER_FILE = "./target/clear_terminal"

[tasks.clean-marker]
script_runner = "@duckscript"
script = '''
rm ${CLEAR_TERMINAL_MARKER_FILE}
'''

[tasks.clear-terminal]
condition = { files_exist = ["${CLEAR_TERMINAL_MARKER_FILE}"] }
# .....

[tasks.enable-clear-terminal]
script_runner = "@duckscript"
script = '''
touch ${CLEAR_TERMINAL_MARKER_FILE}
'''

[tasks.run]
command = "cargo"
args = ["run"]

[tasks.watch-run]
run_task = { name = ["clear-terminal", "enable-clear-terminal", "run"] }
watch = true

[tasks.watch-run-flow]
dependencies = ["clean-marker"]
run_task = "watch-run"
```

its a workaround for this limitation, but couldn't think of a good simple way of communicating env changes from child to parent which is covering all cases.

**Comment by d4h0** [comment]
Thank you, this works!

So if this isn't fixable, I think, a short note within the 'watch' option docs (that warns about this, and mentions this workaround) would be good.

Thanks again!

**Comment by sagiegurari** [comment]
good idea!

**Comment by d4h0** [comment]
I just realized, `cargo watch` actually has a `--clear` option – but it seems, `cargo make` doesn't support this option.

Is there a reason why `cargo make` doesn't support all `cargo watch` options?

Also, I'm wondering what the benefit of using `cargo make`s `cargo watch` binding vs. using `cargo watch` directly in a task is. Is there any disadvantage with the latter?

**Comment by sagiegurari** [comment]
* i am not sure if that option existed when i wrote the integration
* you can do both. but having a task and just adding watch=true makes it simpler usually. but you can do both. you can also call raw cargo watch from cargo make but thats even more verbose.

**Comment by d4h0** [comment]
That makes sense, thanks!

I think, I'll give using cargo watch directly a try. That should be less verbose than the workaround (if I'm not missing anything).
